### PR TITLE
Merge AdditionalLibraryIgnoresMatcher with GlobalIgnoresMatcher

### PR DESCRIPTION
**Pull request by mcculls**
Extracted out from a larger refactoring of our ByteBuddy strategies - combining these matchers and making it a singleton makes it easier to re-use in the redefinition strategy. It also brings a slight reduction in startup time.

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/AdditionalLibraryIgnoresMatcher.java [hunk 1]
```diff
@@ -26,292 +26,327 @@
...
-            || name.startsWith(
-                "org.springframework.beans.factory.groovy.GroovyBeanDefinitionReader$")
-            || name.equals("org.springframework.beans.factory.support.AbstractBeanFactory")
-            || name.equals(
-                "org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory")
-            || name.equals(
-                "org.springframework.beans.factory.support.DefaultListableBeanFactory")) {
-          return false;
+    switch (name.charAt(0) - 'a') {
+        // starting at zero to get a tableswitch from javac, though it looks horrendous
```
Comment (richardstartin) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/AdditionalLibraryIgnoresMatcher.java:30]: I concur

**Comment by richardstartin** [comment]
How much improvement do we get from this? I haven't seen `AdditionalLibraryIgnoresMatcher` come up in any profiles, but that might be a problem with the profiles.

**Comment by richardstartin** [comment]
Another thought, if this makes a difference, what do we get from the separation of `GlobalIgnoresMatcher` and `AdditionalIgnoresMatcher`? Can't the latter be inlined in to the former to share more filtering? These classes will be exercised very early on before we will benefit from any inlining, and look at them, neither is "clean code" anyway.

**Comment by bantonsson** [comment]
According to the docs it's to allow for better testing. I don't know why though, since the `AdditionalIgnoresMatcher` is always _enabled_ in the agent, presumably to speed things up. I think we should merge them as a first step.

```
/**
 * Additional global matchers that are used to reduce number of classes we try to apply expensive
 * matchers to.
 *
 * <p>This is separated from {@link GlobalIgnoresMatcher} to allow for better testing. The idea is
 * that we should be able to remove this matcher from the agent and all tests should still pass.
 * Moreover, no classes matched by this matcher should be modified during test run.
 */
```

**Comment by mcculls** [comment]
I'm +1 for merging them... as you say the doc says this is for testing purposes, but IMHO it's better to test what we ship rather than configurations which never apply in production.

**Review by richardstartin — APPROVED** [review]

**Review by bantonsson — APPROVED** [review]

**Comment by richardstartin** [comment]
How much improvement comes from this and how is it being measured? I found no effect when loading 10000 classes from a dropwizard application. Though may be that the annotation matching is drowning out the signal.

**Comment by richardstartin** [comment]
(I would accept this change on the basis of it being an improvement.)

**Comment by mcculls** [comment]
@richardstartin when the PR was just introducing the tableswitch to `AdditionalLibraryIgnoresMatcher` I saw a small improvement, but it was wasn't statistically significant given the variability in startup time. After merging the two classes I no longer see that difference.

But I still think merging them is reasonable from a code cleanup point-of-view.

**Comment by mcculls** [comment]
I'm going to look at tackling this another way with an aim to reduce the amount of bytecode

**Pull request closed without merging**